Making a way to order IQueryable objects with a custom comparer.

peter-majeed opened the pull request:
The motivation for this PR was the question @ [Naturally sort IQueryable strings in WCF](http://stackoverflow.com/questions/43565360/naturally-sort-iqueryable-strings-in-wcf) that I mentioned several weeks ago. This seemed like the best way to handle OData handling of IQueryable objects in memory for both WCF and WebAPI.

Ideally, this would be open-sourced and put on nuget.org. There's some potential play with TravisCI to do automated testing, etc., but I figured that could go in a separate PR.

Commit:
Putting corefx classes in project namespaces.

This will make it easier to distinguish between corefx code and
project-related code.

Commit:
Removing unnecessary extension methods.

These methods are better served by the functionality in corefx.

Commit:
Extending EnumerableQuery functionality to accept IComparers.

Introduces the notion of an IComparer to the EnumerableQuery to allow
for ordering with default comparers.

Commit:
Including functionality to naturally sort strings.

We *could* make this functionality in another package, but it's
relevant to the problem this repo is trying to solve. So it seemed
relevant.

Commit:
Grouping projects in their position within the folder structure.

Commit:
Testing the NaturalSort algorithm.

Commit:
Testing the NaturalQueryable flow.

Commit:
Adding licenses and documentation.

Commit:
Adding nuspec packages to the projects.

Commit:
Merge branch 'importcorefx-queryable' into initialrelease

peter-majeed commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
I figured we'd go with the MIT License for this project. It's what we're using for our open source FE packages as well.

peter-majeed commented on LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2017 Buildium, LLC
```
Some of our FE packages copyright as **Buildium**, but it seemed more accurate to use **Buildium, LLC** as our official name.

peter-majeed commented on src/ComparedQueryable/ComparedQueryable.csproj:
```diff
@@ -40,15 +40,22 @@
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="NativeQueryable\CachedReflection.cs" />
+    <Compile Include="EnumerableExecutor[TCollectionElement,TComparison].cs" />
```
I wanted a way to include generic classes as their own files. There was a naming convention @ [Convention for Filenames of Generic Classes](http://stackoverflow.com/a/1179581/684271) that I took a liking to here.

Doelfke commented on test/NaturalSort.Test/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("NaturalSort.Test")]
```
Should any of this be filled in?

peter-majeed replied:
I didn't think it was important to fill in the assembly info for the test projects.

peter-majeed commented on src/ComparedQueryable/ComparedQueryable.csproj:
```diff
@@ -40,15 +40,22 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\NaturalSort\NaturalSort.csproj">
```
Ideally, we would first put out the nuget package for the NaturalSort algorithm, then import that package, but for now, this helps gets things working.

peter-majeed commented on src/ComparedQueryable/ComparedQueryable.nuspec:
```diff
@@ -0,0 +1,21 @@
+<package >
+  <metadata>
+    <id>Buildium.ComparedQueryable</id>
+    <version>0.1.0</version>
```
I made this first version of the `ComparedQueryable` project version 0.1.0 to indicate that it was in a very beta state.

peter-majeed commented on src/ComparedQueryable/ComparedQueryable.nuspec:
```diff
@@ -0,0 +1,21 @@
+    <id>Buildium.ComparedQueryable</id>
+    <version>0.1.0</version>
+    <authors>Peter A Majeed</authors>
+    <owners>Buildium, LLC</owners>
```
Is this the right owner for this project?

peter-majeed commented on src/ComparedQueryable/EnumerableQuery[TCollectionElement,TComparison].cs:
```diff
@@ -0,0 +1,49 @@
+
+        internal override EnumerableRewriter GetEnumerableRewriter()
+        {
+            return new EnumerableRewriter<TComparison>(m_comparer);
```
The `EnumerableRewriter` is what does most of the heavy lifting of the `IQueryable` operations - we just need a way to create one here while passing it an `IComparer<T>`.

peter-majeed commented on src/ComparedQueryable/EnumerableRewriter[TComparison].cs:
```diff
@@ -0,0 +1,77 @@
+        private readonly IComparer<TComparison> m_comparer;
+        private static readonly ISet<string> ComparerFunctions = new HashSet<string>(new[]
+        {
+            nameof(Queryable.OrderBy),
```
There are only four methods that accept an `IComparer<T>` for `IQueryable<T>` - predictably, the four ordering methods.

peter-majeed commented on src/ComparedQueryable/EnumerableRewriter[TComparison].cs:
```diff
@@ -0,0 +1,77 @@
+            {
+                return queryableArgs;
+            }
+            return queryableArgs.Concat(new[] { Expression.Constant(m_comparer) }).ToList().AsReadOnly();
```
This is the line of code that allows us to use the `IComparer` as the default second argument of the `OrderBy` clauses.

peter-majeed commented on src/ComparedQueryable/EnumerableRewriter[TComparison].cs:
```diff
@@ -0,0 +1,77 @@
+            }
+
+            var methodInfoWithComparer = potentialMethodMatches
+                .FirstOrDefault(MethodInfoHasIComparerParameter);
```
This is some extra insurance to make sure that the method that we're calling on the `IQueryable` accepts an `IComparer<T>` argument.

peter-majeed commented on src/ComparedQueryable/LICENSE:
```diff
@@ -0,0 +1,23 @@
+The MIT License (MIT)
```
This license was copied verbatim from the .NET corefx project root.

peter-majeed commented on src/ComparedQueryable/NativeQueryable/EnumerableExecutor.cs:
```diff
@@ -34,11 +35,17 @@ public EnumerableExecutor(Expression expression)
             return func();
         }
+
+        internal virtual EnumerableRewriter GetEnumerableRewriter()
```
Had to extend these classes to allow for the generic subclasses to be able to extend its functionality.

peter-majeed commented on src/ComparedQueryable/NativeQueryable/Error.cs:
```diff
@@ -2,28 +2,31 @@
 
         internal static Exception ArgumentNotIEnumerableGeneric(string message) =>
-            new ArgumentException(Strings.ArgumentNotIEnumerableGeneric(message));
+            new ArgumentException(message);
```
There were some shared messages that corefx had here that I didn't think necessarily needed to be built out for error tracking in this project.

peter-majeed commented on src/ComparedQueryable/QueryableExtensions.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <param name="collection"></param>
+        /// <param name="comparer">Used for ordering operations.</param>
+        /// <returns></returns>
+        public static IQueryable<T> AsComparedQueryable<T, TComparison>(this IEnumerable<T> collection,
```
This method is the generic method that, when called on an `IQueryable`, orders any sub-properties that match the type of the passed type of the `IComparer<T>`.

peter-majeed commented on src/ComparedQueryable/QueryableExtensions.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <typeparam name="T"></typeparam>
+        /// <param name="collection"></param>
+        /// <returns></returns>
+        public static IQueryable<T> AsNaturalQueryable<T>(this IEnumerable<T> collection)
```
This is a practical application of the generic extension method below that sorts strings naturally (the point of this PR).

peter-majeed commented on src/NativeQueryable/Strings.cs:
```diff
@@ -1,26 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
This file didn't seem terribly useful, so I removed it.

peter-majeed commented on src/NaturalSort/NaturalSort.csproj:
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
I made the NaturalSort project a .NET standard project since all it was doing was string manipulation/ordering.

peter-majeed commented on src/NaturalSort/NaturalSort.nuspec:
```diff
@@ -0,0 +1,19 @@
+<package >
+  <metadata>
+    <id>Buildium.NaturalSort</id>
+    <version>1.0.0</version>
```
This project seemed to make sense to have a version 1.0.0, since its functionality was already tested with Buildium.

peter-majeed commented on src/NaturalSort/NaturalSort.nuspec:
```diff
@@ -0,0 +1,19 @@
+  <metadata>
+    <id>Buildium.NaturalSort</id>
+    <version>1.0.0</version>
+    <title>NaturalSortComparer</title>
```
In this case, the Id of the project is `Buildium.NaturalSort`, but the name it would show up with in nuget.org would be `NaturalSortComparer`. That got me to thinking - how do we want to name our packages in Nuget? Using .NET namespacing with Buildium at the front, or with more functional names?

peter-majeed commented on src/NaturalSort/NaturalSort.nuspec:
```diff
@@ -0,0 +1,19 @@
+    </description>
+    <releaseNotes>Initial release of the natural string comparer.</releaseNotes>
+    <copyright>Copyright 2017 Buildium, LLC</copyright>
+    <tags>natural sort comparer icomparer order orderby string</tags>
```
Seemed like reasonable search terms in Nuget.

peter-majeed commented on src/NaturalSort/NaturalSortComparer.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        /// <inheritdoc />
+        public int Compare(string x, string y)
```
This entire function came from Buildium.

peter-majeed commented on src/NaturalSort/NaturalSortComparer.cs:
```diff
@@ -0,0 +1,119 @@
+        {
+        }
+
+        private NaturalSortComparer()
```
It didn't seem to make sense to be able to new up instances of this comparer, since it doesn't maintain any state, so I made it non-constructable and provided a static instance.

peter-majeed commented on test/ComparedQueryable.Test/packages.config:
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="FluentAssertions" version="4.19.2" targetFramework="net462" />
+  <package id="xunit" version="2.2.0" targetFramework="net462" />
```
xunit seemed like a nice testing framework to use. Plus it seemed compatible with .NET core projects.

peter-majeed reviewed (commented):
Leaving some hopefully useful guiding comments.

peter-majeed commented:
@herkyg primary?

bernard-chen commented on test/ComparedQueryable.Test/ComparedQueryableTest.cs:
```diff
@@ -0,0 +1,107 @@
+            // GIVEN I have a collection with a string property
+            var twoNorthStreet = new Address("2 N. Street");
+            var tenNorthStreet = new Address("10 N. Street");
+            var collection = new[] { twoNorthStreet, tenNorthStreet };
```
I think it's worth having a test where the initial collection has tenNorthStreet first so that we prove that we're actually sorting. Otherwise, the implementation could just return any results in the same order they were originally given.

bernard-chen commented on test/ComparedQueryable.Test/ComparedQueryableTest.cs:
```diff
@@ -0,0 +1,107 @@
+
+namespace ComparedQueryable.Test
+{
+    public class ComparedQueryableTest
```
Can we have another test where the numbers that distinguish between the values are not at the beginning of the strings? It might be "Option 2" vs "Option 10".

herkyg replied:
This is a good suggestion, although I don't know if there's a clear answer for what the desired behavior should be. The first thing that came to mind for me was similar to yours: "123 N 2nd St" vs "123 N 10th St". But what about "S/N 5EFFH" vs "S/N 15AAA"? I would want 15AAA to come before 5EFFH. So perhaps it's a word boundary that defines the appropriate sorting behavior? Maybe we only sort numbers that come at the beginning of the string? Although I don't think that's going to be satisfactory (consider "Unit 1", etc.)
But: additional tests will help document what we define as "correct" behavior in these scenarios.

peter-majeed replied:
I didn't want to test this given that we have a separate test for the natural sort test algorithm. Otherwise, we end repeating the same test cases in both cases.

Commit:
Making sure that tests don't make too many assumptions.

Reordering our test data to make sure that the system doesn't pretend to
naively return items in a collection in the order that the items were in
the original collection.

herkyg commented on src/ComparedQueryable/NativeQueryable/EnumerableRewriter.cs:
```diff
@@ -54,13 +56,13 @@ protected override Expression VisitMethodCall(MethodCallExpression m)
             {
                 List<Expression> newArgs = null;
-                for (int i = 0, n = pis.Length; i < n; i++)
+                for (int i = 0, n = argList.Count; i < n; i++)
```
I think here pis.Length should be equal to argList.Count? Why make the change in line 63/65? If the lengths are the same, it really doesn't matter, but you're checking one length and then looping based on another, which is unexpected and doesn't seem to improve the code.

peter-majeed replied:
I changed this back. There was a version where I had been playing around with this method, but it's not needed once the method is made virtual.

peter-majeed replied:
@herkyg actually, I changed this back again. I got confused - this is actually the crux of the change, though this time, I left more comments in the code explaining why these choices were made.

Commit:
Reverting method behavior change.

This method was unintentionally added.

Commit:
Removing nuspec file for .NET standard project.

.NET standard projects accept package information in the .csproj file
rather than a separate .nuspec file.

Commit:
Declaring NaturalSort as a NuGet dependency.

Since NaturalSort is on NuGet, let's import the project as a NuGet
dependency rather than a project reference.

This project reference will likely live in a separate repository at
some point so that a user interested in the natural sort algorithm
wouldn't have to download the IQueryable project as well.

Commit:
Making it clearer why we alter the EnumerableRewriter.

Let's make sure we're able to modify the EnumerableRewriter correctly
when passing in a custom comparer.

Commit:
Testing special characters when sorting.

Special characters should also be naturally sorted.

peter-majeed commented:
@herkyg could you have a final pass at these changes? I'm not sure if there's a good way in a Github PR to approve, but let me know if you're ready to move forward.

peter-majeed commented:
Merged with master.
